packages/next/types/index.d.ts

```diff
@@ -65,17 +65,30 @@ export {
   NextApiHandler,
 }
 
-export type GetStaticProps<
-  P extends { [key: string]: any } = { [key: string]: any },
-  Q extends ParsedUrlQuery = ParsedUrlQuery
-> = (ctx: {
+export type GetStaticPropsContext<Q extends ParsedUrlQuery = ParsedUrlQuery> = {
   params?: Q
   preview?: boolean
   previewData?: any
-}) => Promise<{
+}
+
+export type GetStaticPropsReturnType<P> = {
   props: P
   unstable_revalidate?: number | boolean
-}>
+}
+
+export type GetStaticProps<
+  P extends { [key: string]: any } = { [key: string]: any },
+  Q extends ParsedUrlQuery = ParsedUrlQuery
+> = (ctx: GetStaticPropsContext<Q>) => Promise<GetStaticPropsReturnType<P>>
+
+export type InferredStaticProps<
```

Comment:
Can we infer the "inferred" behavior here?
```suggestion
export type StaticProps<
```